### ISSUE TITLE
Custom callback_path uses bait and switch with request_path

### ISSUE THREAD

**Issue by canadaduane** [omniauth/omniauth#630]
It looks like a copy-paste error to me:

https://github.com/intridea/omniauth/blob/master/lib/omniauth/strategy.rb#L378

I expect to be able to write a custom callback_path such as:

``` ruby
provider :singly, ID, SECRET_ID, :callback_path => Proc.new{ |env| "/my/path/#{env[:key]}/callback" }
```

But line 378 uses request_path instead of callback_path for callables.

**Comment by tmilewski** [comment]
Closing per #693

**Comment by emirikol** [comment]
Was this ever addressed? it doesn't seem like #693 was actually merged.

**Comment by leo-souza** [comment]
I was facing issues with having a dynamic `callback_path`, so this is how we solved it (seems bit hackish but I couldn't find any other way)
Our case: We have 2 authentications: one for User (login) and another for Profile (to grab page data).
```ruby
provider :facebook, ENV["FACEBOOK_APP_ID"], ENV["FACEBOOK_APP_SECRET"], image_size: 'large',
    scope: '...our scope..',
    info_fields: '..our info fields...',
    callback_path: ->(env) do
      env['REQUEST_PATH'].to_s.match(/\/auth\/facebook\/callback/)
    end,
    setup: ->(env) do
      req = Rack::Request.new(env)
      env['omniauth.strategy'].options[:callback_path] = "/#{req.params['resource'] == 'user' ? 'users' : 'profiles'}/auth/facebook/callback"
    end

on_failure do |env|
  if env['omniauth.params'].to_h['resource'] == 'user'
    env["devise.mapping"] = Devise.mappings[:user]
    Users::OmniauthCallbacksController.action(:failure).call(env)
 else
    env["devise.mapping"] = Devise.mappings[:profile]
    Profiles::OmniauthCallbacksController.action(:failure).call(env)
  end
end
```
so for profiles our url is `/auth/facebook?scope=<different scope>` and for users is `/auth/facebook?resource=user`
and we split the call routes like this
```ruby
devise_scope :profile do
  get '/profiles/auth/facebook/callback', to: 'profiles/omniauth_callbacks#facebook'
end

devise_scope :user do
  get '/users/auth/facebook/callback', to: 'users/omniauth_callbacks#facebook'
end
```
